### PR TITLE
Doc Reorg: Unpackaged Metadata & Metadata ETL

### PR DESCRIPTION
**Pull request by davidmreed**
Rendered:
https://cumulusci.readthedocs.io/en/feature-doc-reorg__unpackaged/unpackaged.html and https://cumulusci.readthedocs.io/en/feature-doc-reorg__unpackaged/metadata_etl.html

# Critical Changes

# Changes

# Issues Closed

### CODE REVIEW

File: docs/metadata_etl.rst [hunk 1]
```diff
@@ -0,0 +1,198 @@
+.. _metadata-etl:
```
Comment (davidmreed) [inline docs/metadata_etl.rst:1]: This is copied verbatim from the current docs. I thought there was too much material to go with `unpackaged`.

File: docs/metadata_etl.rst [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+``BaseMetadataTransformTask`` can be used as the base class for ETL tasks that require
+more flexibility than is permitted by ``MetadataSingleEntityTransformTask``, such as tasks
+that must mutate multiple Metadata API entities in a single operation. Subclasses must
+override ``_get_entities()`` to return a dict mapping Metadata API entities to collections of
+API names. (The base class will generate a corresponding ``package.xml``). Subclasses must
+also implement ``_transform()``, as with ``BaseMetadataETLTask``.
+
+``UpdateFirstAttributeTextTask`` is a base class and generic concrete task that makes it easy to
+perform a specific, common transformation: setting the value of a the first instance of a specific
```
Comment (Br4nd0R) [inline docs/metadata_etl.rst:183]:
```suggestion
perform a specific, common transformation: setting the value of the first instance of a specific 
```

**Review by Br4nd0R — CHANGES_REQUESTED** [review]
Just a minor tweak on the bottom there. Great write-up, I learned lots :)

File: docs/metadata_etl.rst (hunk 1 above)
Comment (davisagli) [reply docs/metadata_etl.rst:1]: I like positioning it right after `unpackaged` and making sure that both topics contextualize themselves in relation to the other. They are two different approaches to the same problem space of post-install configuration.

File: docs/unpackaged.rst [hunk 3]
```diff
@@ -1,14 +1,161 @@
...
+
+For example, a testing-oriented scratch org may wish to deploy a customized set of Page Layouts to help testers easily visualize data under test. Such page layouts could be stored in ``unpackaged/config/qa``.
+
+
+Unpackaged Metadata Folder Structure
+------------------------------------
+
+All unpackaged metadata is stored in the ``unpackaged`` directory tree. Within this directory are the three top-level directories, ``unpackaged/pre``, ``unpackaged/post``, and ``unpackaged/config``.
+
+These trees contain metadata bundles in Metadata API format: that is, a directory containing a ``package.xml`` manifest and Metadata API-format source code. CumulusCI does not support Salesforce DX format for unpackaged bundles.
```
Comment (davisagli) [inline docs/unpackaged.rst:52]: Actually since the merge of the 2gp branch, I believe if you point `deploy` at a folder that doesn't have a package.xml, it will use sfdx to convert it in a tempdir to deploy. But this may not work well with namespace tokens in some locations if they get in the way of the conversion. There is potential to make it work with a scheme like this:
1. Copy sfdx format to a tempdir.
2. Do namespace injection.
3. Convert to mdapi.

File: docs/unpackaged.rst [hunk 4]
```diff
@@ -1,14 +1,161 @@
...
+
+.. _namespace-injection:
+
+Namespace Injection
 -------------------
 
-Customize config flows
+Projects that build managed packages often must construct their unpackaged metadata to be deployable in multiple contexts: unmanaged deployments, such as developer orgs, unmanaged namespaced scratch orgs, and also managed contexts, such as a beta test org or a demo org created with ``install_prod``.
+
+  Projects that are building an org implementation (not a package with a namespace) do not need to use namespace injection.
```
Comment (davisagli) [inline docs/unpackaged.rst:61]: You haven't defined what namespace injection is yet...

File: docs/unpackaged.rst [hunk 5]
```diff
@@ -1,14 +1,161 @@
...
+The task ``deploy_post``, which is part of the ``config_dev``, ``config_qa``, and ``config_managed`` flows, is responsible for deploying these bundles.
+
+Metadata that's not intended to be delivered to all installations of the product should *not* be included in ``unpackaged/post``. It's critical for managed package projects that this metadata include namespace tokens (see :ref:`namespace-injection`).
+
+``unpackaged/config``: Tailoring Orgs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Projects may come with more than one supported configuration in their CumulusCI automation. For example, projects may support distinct, tailored ``dev_org``, ``qa_org``, and ``install_prod`` flows, each of which performs different setup for their specific use case. Unpackaged metadata stored in ``unpackaged/config`` is a tool to support operational needs to tailor orgs to different configurations. 
+
+For example, a testing-oriented scratch org may wish to deploy a customized set of Page Layouts to help testers easily visualize data under test. Such page layouts could be stored in ``unpackaged/config/qa``.
```
[the remark this answers is not in the document]
Comment (davisagli) [reply docs/unpackaged.rst:44]: Oh, I see that's what is in the "Customize Config Flows" section below.

**Review by davisagli — CHANGES_REQUESTED** [review]
Nice work.

File: docs/metadata_etl.rst [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+ - ``assign_compact_layout`` assigns a Compact Layout to a given sObject.
+ - ``set_duplicate_rule_status`` activates and deactivates Duplicate Rules.
+ - ``set_field_help_text`` sets the Help Text of a given field (a non-push-upgradeable attribute).
+ - ``set_organization_wide_defaults`` sets the Org-Wide Defaults for one or more sObjects,
+   and waits until sharing recalculation completes before proceeding.
+
+Most Metadata ETL tasks accept the option ``api_names``, which specifies the scope of the operation.
+In most cases, more than one entity may be transformed in a single operation. Each task is a single
+atomic Metadata API deployment operation. Consult the Task Reference or use the ``cci task info``
+command for more information on the usage of each task.
```
[the remark this answers is not in the document]
Comment (davidmreed) [reply docs/metadata_etl.rst]: I revised to be more clear. Do you think there's a need for an explicit disclaimer?

[the remark this answers is not in the document]
Comment (davisagli) [reply docs/metadata_etl.rst]: Not sure... I think it's a bit incongruent with how we're emphasizing the safety of the Metadata ETL approach. On the other hand, it's pretty unlikely to occur. But it's probably worth mentioning in a note callout as fine print, so there's something we can point to if someone tries to run multiple concurrent operations and has trouble.

File: docs/metadata_etl.rst [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+Namespace Injection
+-------------------
+
+All out-of-the-box Metadata ETL tasks accept a Boolean ``managed`` option. If ``True``, CumulusCI
+will replace the token ``%%%NAMESPACE%%%`` in API names and in values used for transforming metadata
+with the project's namespace; if ``False``, the token will simply be removed.
+
+
+Implementation of Metadata ETL Tasks
```
[the remark this answers is not in the document]
Comment (davidmreed) [reply docs/metadata_etl.rst]: I moved it to a new PR, https://github.com/SFDO-Tooling/CumulusCI/pull/2101

File: docs/metadata_etl.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidmreed) [reply docs/metadata_etl.rst]: That's a good point. I added a note to that effect.

File: docs/unpackaged.rst [hunk 8]
```diff
@@ -1,14 +1,212 @@
...
+Not everything that's part of an application can be part of a package.
 
-Namespace injection
+CumulusCI implements the Product Delivery Model by offering support for complex applications -- applications that may include multiple managed packages as well as unpackaged metadata, and setup automation that configures org settings or makes precise changes to existing configuration.
+
+The tools used to implement that support are *unpackaged metadata* and *Metadata ETL*. 
+
+Unpackaged metadata refers to metadata that is not delivered as part of a package, and can include both support metadata delivered to users as well as metadata that operationally configures orgs used by the product. 
+
+Metadata ETL is a suite of tasks that support surgically altering existing metadata in an org. It's a powerful technique that changes the unpackaged configuration in an org without risking damage by overwriting existing customizations with stored metadata. Metadata ETL is relevant for delivering applications to customers safely, and is often a superior alternative to unpackaged metadata.
```
Comment (justinvholt) [inline docs/unpackaged.rst:12]: Would we call it a feature rather than a technique? And the end of this sentence is a bit confusing: Do the existing customizations have stored metadata? Or does Metadata ETL use stored metadata to not damage existing customizations? (I’m  95% certain it’s the former, but I would like confirmation.)

Comment (justinvholt) [inline docs/unpackaged.rst:12]: Would we call it a feature rather than a technique? And the end of this sentence is a bit confusing: Do the existing customizations have stored metadata? Or does Metadata ETL use stored metadata to not damage existing customizations? (I’m 95% certain it’s the former, but I would like confirmation.)

File: docs/unpackaged.rst [hunk 9]
```diff
@@ -1,14 +1,212 @@
...
+
+The ``deploy_pre`` task, which is part of the ``dependencies`` flow, is responsible for deploying these bundles.
+
+.. important:: Do not include metadata in ``unpackaged/pre`` unless it is intended to be delivered to *all* installations of the product.
+
+
+``unpackaged/post``: Configuration After Package Install
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Projects often include metadata that is genuinely part of the application, but either cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the project's own code and metadata are deployed first and the org is configured.
```
Comment (justinvholt) [inline docs/unpackaged.rst:39]: Changed up some of the wording in the first sentence and added the second one. Please confirm that everything is correct.

File: docs/unpackaged.rst [hunk 10]
```diff
@@ -1,14 +1,212 @@
...
 -------------------
 
-Customize config flows
+Projects that build managed packages often construct their unpackaged metadata to be deployable in multiple contexts.
+
+* Unmanaged deployments, such as developer orgs
+* Unmanaged namespaced scratch orgs
+* Managed contexts, such as a beta test org or a demo org created with ``install_prod``
+
+Because projects that are building an org implementation or a non-namespaced package do not have a namespace, or a distinction between managed and unmanaged contexts, these projects must also build namespaced packages.
```
Comment (justinvholt) [inline docs/unpackaged.rst:91]: Why must they build namespaced packages? I would like to cap this sentence with just a few words.

"These projects must also build namespaced packages to __________________."

File: docs/unpackaged.rst [hunk 11]
```diff
@@ -1,14 +1,212 @@
...
+
+    Example: Metadata located in ``unpackaged/post`` is deployed after the application code in both unmanaged and managed contexts. If that metadata contains references to the application components, it must be deployable when that metadata is nampespaced (in a managed context or namespaced scratch org) *and* when it is not (in an unmanaged context).
```
Comment (justinvholt) [inline docs/unpackaged.rst:93]: This is getting hard to follow simply because I don’t fully know why things are required to be namespaced. Which is okay because I'm still quite the novice when it comes to building anything with ``cci``. That said, is it safe to presume that most devs this deep into the material know about namespaces, why they're needed and will follow this material with (relative) ease?

File: docs/unpackaged.rst [hunk 12]
```diff
@@ -1,14 +1,212 @@
...
+
+CumulusCI uses a strategy called *namespace injection* to support this use case. Namespace injection is very powerful, and requires care from application implementors to ensure that metadata remains deployable in all contexts.
```
Comment (justinvholt) [inline docs/unpackaged.rst:95]: "application implementors"

Just confirming this is what we want to call them because this is the first time I've heard this label being used.

File: docs/unpackaged.rst [hunk 13]
```diff
@@ -1,14 +1,212 @@
...
+* ``%%%NAMESPACED_ORG_OR_C%%%`` is replaced with the package's namespace in a namespaced org *only*, not in a managed installation. Otherwise, it is replaced with ``c``, the generic namespace used in Lightning components.
+* ``%%%NAMESPACE_DOT%%%`` is replaced with the package’s namespace in any context with a namespace (such as a namespaced org or managed org) followed by a period (``.``) rather than two underscores.
+    .. note:: This token is used to construct references to packaged Record Types.
+
+..
+
+    Example: A portion of metadata from the Nonprofit Success Pack (NPSP) is stored in a subdirectory under ``unpackaged/post``, meaning it's deployed after the application metadata. This metadata updates a Compact Layout on the ``Account`` object, and references packaged metadata from the application as well as from other managed packages. To complete these tasks, this metadata requires the use of namespace tokens to represent the ``npsp`` namespace, letting CumulusCI automatically adapt the metadata to deploy into managed and unmanaged contexts.
```
Comment (justinvholt) [inline docs/unpackaged.rst:109]: “To complete these tasks...”

Added this for clarity, so I’m simply confirming that the updating and referencing are indeed tasks. And if not, is there a better word? Steps?

File: docs/unpackaged.rst [hunk 14]
```diff
@@ -1,14 +1,212 @@
...
+            <fields>npo02__TotalOppAmount__c</fields>
+            <fields>%%%NAMESPACE%%%Number_of_Household_Members__c</fields>
+            <label>NPSP Household Account</label>
+        </compactLayouts>
+    </CustomObject>
+
+..
+
+    Note that only the reference to the NPSP field ``Number_of_Household_Members__c`` is tokenized. (This field is called ``npsp__Number_of_Household_Members__c`` when installed as part of the managed package.) References to NPSP's own managed package dependency, ``npo02``, are not tokenized because this metadata is always namespaced when installed.
+
```
Comment (justinvholt) [inline docs/unpackaged.rst:127]: Why this specific field? Also, is this detail in parentheses relevant/helpful for a dev to know for this example to make sense? I honestly can't make heads or tails of it.  To be honest, it feels like we’re currently missing one or two details here about why only this field is the one to get tokenized.

File: docs/unpackaged.rst [hunk 15]
```diff
@@ -1,14 +1,212 @@
...
+    If this metadata isn't tokenized, it fails to deploy into an org containing NPSP as a beta or released managed package.
+
```
Comment (justinvholt) [inline docs/unpackaged.rst:129]: Why? This example might need more clarification, or it might need to be swapped out with one easier to comprehend.

(Or it might be completely fine for an advanced dev. Your call.)

File: docs/unpackaged.rst [hunk 16]
```diff
@@ -1,14 +1,212 @@
...
+.. note:: The resolution of component references in namespaced scratch orgs and in managed installations of the same metadata are not identical. Metadata that is tokenized and deploys cleanly in a namespaced scratch org can still fail in a managed context.
```
Comment (justinvholt) [inline docs/unpackaged.rst:130]: This seems like real inside baseball. Is it reasonable to presume that advanced devs will understand this? (Again, it's completely fine if I don't because I presume at this point that only advanced devs would be this far into the written materials.)

File: docs/unpackaged.rst [hunk 17]
```diff
@@ -1,14 +1,212 @@
...
+* Whenever possible, build your unpackaged metadata in an org that contains a beta or released managed package. By doing so, the metadata contains namespaces when extracted, which CumulusCI easily replaces with tokens when retrieving metadata. It's difficult to manually tokenize metadata that's retrieved from an unmanaged org without namespaces. 
+
+After building changes to unpackaged metadata in a managed org, retrieve it using ``retrieve_changes`` with the additional ``namespace_tokenize`` option, and use the ``path`` option to direct the retrieved metadata to your desired unpackaged directory.
+
+    Example: Run ``retrieve_changes`` to retrieve metadata changes into the ``unpackaged/config/qa`` subdirectory, and replace references to the namespace ``npsp`` with the appropriate token.
+
+.. code-block:: console
+
+    $ cci task run retrieve_changes --path unpackaged/config/qa --namespace_tokenize npsp
```
Comment (justinvholt) [inline docs/unpackaged.rst:151]: Swapped out the old -o flags. Please confirm this updated command is correct.

File: docs/unpackaged.rst (hunk 8 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:12]: [ETL](https://en.wikipedia.org/wiki/Extract,_transform,_load) is an established technique/procedure/methodology. I wouldn't refer to it as a feature.

Ya, this sentence could be better. Maybe: "It's a powerful technique that alters the unpackaged configuration in an org without risking damage to existing customization by overwriting them with incoming metadata."

Let me know if that's any better.

File: docs/unpackaged.rst (hunk 9 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:39]:
```suggestion
Projects often include metadata that is genuinely part of the application, but either cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the package's own code and metadata are deployed first and the org is configured.
```

File: docs/unpackaged.rst (hunk 10 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:91]: @davidmreed - Is a non-namespaced package just another way of saying 'unamangeg/unlocked' package? Also wondering about Justin's question here. The following example outlines how `unpackaged/post` needs to work in both managed and unmanaged contexts; however, if I'm building an org implementation in a package without a namespace, what's the reason for needing to build namespaced packages?

File: docs/unpackaged.rst (hunk 11 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:93]: This example is clear to me (not sure how helpful this is), but I think it should be safe to assume that people reading this are going to be interested in, or familiar with, Salesforce packages in general. We might want to change this example based on feedback from David R. regarding your prior question.

File: docs/unpackaged.rst (hunk 12 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:95]: We could maybe use "application developers and architects" as a substitute?

File: docs/unpackaged.rst (hunk 8 above)
Comment (davidmreed) [reply docs/unpackaged.rst:12]: Yes to the Brandon edit!

File: docs/unpackaged.rst (hunk 9 above)
Comment (davidmreed) [inline docs/unpackaged.rst:39]:
```suggestion
Projects often include metadata that is genuinely part of the application, but cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the project's own code and metadata are deployed first and the org is configured.
```

File: docs/unpackaged.rst (hunk 10 above)
Comment (davidmreed) [reply docs/unpackaged.rst:91]: I think I wrote this sentence, but it's total nonsense. 

```suggestion
Because projects that are building an org implementation or a non-namespaced package do not have a namespace, or a distinction between managed and unmanaged contexts, these projects typically don't need to use namespace injection.
```

File: docs/unpackaged.rst (hunk 11 above)
Comment (davidmreed) [inline docs/unpackaged.rst:93]:
```suggestion
    Example: Metadata located in ``unpackaged/post`` is deployed after the application code in both unmanaged and managed contexts. If that metadata contains references to the application components, it must be deployable when that metadata is namespaced (in a managed context or namespaced scratch org) *and* when it is not (in an unmanaged context).
```

Comment (davidmreed) [reply docs/unpackaged.rst:93]: I'm happy to expand on or clarify this example but I think it's an important one because it is very common.

File: docs/unpackaged.rst (hunk 12 above)
Comment (davidmreed) [reply docs/unpackaged.rst:95]:
```suggestion
CumulusCI uses a strategy called *namespace injection* to support this use case. Namespace injection is very powerful, and requires care from the project team to ensure that metadata remains deployable in all contexts.
```

File: docs/unpackaged.rst (hunk 13 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:109]: Hm. Not sure that we need that there. The "update" and "referencing" in the paragraph are referring to characteristics of particular piece of metadata, not things that will be done in the future. When deployed, the metadata updates a Compact Layout on the Account object, and it also has references to another managed package (in the form of the `npo02__TotalOppAmount__c` field).

Comment (davidmreed) [reply docs/unpackaged.rst:109]: "Steps" would be fine. "Tasks" is definitely confusing and ambiguous.

File: docs/unpackaged.rst (hunk 14 above)
Comment (davidmreed) [reply docs/unpackaged.rst:127]: There's no specific importance to this field, it's just a good example.

But if it's not clear why this field is tokenized, the explanation so far has failed. Should we talk through it? n.b. this are is _notoriously_ difficult to understand for users and even release engineers.

File: docs/unpackaged.rst (hunk 15 above)
Comment (davidmreed) [reply docs/unpackaged.rst:129]:
```suggestion
    If this metadata isn't tokenized, it fails to deploy into an org containing NPSP as a beta or released managed package — because in that context, the field `Number_of_Household_Members__c` is namespaced, as `npsp__ Number_of_Household_Members__c`, and we must refer to it as such.

```

File: docs/unpackaged.rst (hunk 16 above)
Comment (davidmreed) [reply docs/unpackaged.rst:130]: This is Deep Magic. I think it's an important note but it's okay that not everyone gets it. Those who are operating with these tools (namespaced scratch orgs + managed packages) will hopefully get it.

File: docs/unpackaged.rst (hunk 17 above)
Comment (davidmreed) [reply docs/unpackaged.rst:151]: This looks correct to me.

File: docs/unpackaged.rst (hunk 14 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:127]: Salesforce has a convention of using `namespace__` in front of the field name to indicate which package a particular field belongs to. This particular field is the only one in the compact layout that is a member of the `npsp` package (this is why it's the only one getting the token). We know this only because we were told so here. 

The detail in the parens does help I believe, it's letting us know that in a "managed" context this field will be display as `nspsp__Number_of_Household_Members__c` in an org. This is useful because then I can see how the `%%%NAMESPACE%%%` token is getting replaced with `npsp__` in a "managed" context.

File: docs/unpackaged.rst (hunk 15 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:129]: We are pretty deep in the weeds here. We should assume that users reading this section are likely a release engineer familiar with packaging basics. But I think we could maybe make this a little more explicit in how the error actually occurs.

@davidmreed - I need you help on how to finish the example. I believe I'm correct up until "the field would be deployed" part. Is the error on deploy, or is this more related with a problem that occurs in the `package.xml` file, or something else?

```suggestion
    If this metadata isn't tokenized, it will fail to deploy into an org containing NPSP as a beta or released managed package. Since all metadata is retrieved in an unmanaged context (i.e. without a namespace prefix) during development, if there is no `%%%NAMESPACE%%%` token present, the field would be deployed without a namespace, and thus result in an error.
```

File: docs/unpackaged.rst (hunk 16 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:130]: I think this should be fine. There's a lot of jargon, but a release engineer or advanced dev is going to be the target role for this content.

File: docs/unpackaged.rst (hunk 17 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:151]: 10-4

File: docs/unpackaged.rst (hunk 11 above)
Comment (Br4nd0R) [reply docs/unpackaged.rst:93]: Let's keep this then.
